### ISSUE TITLE
onActivate: Could not find a coercion from type java.lang.String to type X

### ISSUE THREAD

**Issue by uklance** [uklance/tapestry-cometd#62]
`onActivate(Foo foo)` is in tapestry-cometd is acting differently to normal tapestry page rendering. It can't not coerce hibenrate entities for example.

The `ArrayEventContext` does not seem to be finding / invoking the `ValueEncoder` in:
`eventContext.get(Class<T> type, in index)`

From debugging, the normal onActivate is invoked with:

```
{org.apache.tapestry5.internal.*URLEventContext*@10311}"EventContext:
   null"
      - valueEncoder = {$ContextValueEncoder_9660c5cdf431@10618}""
      - values = {java.lang.String[1]@10619}
      - values = null
```

Whereas tapestry-cometd is using:

```
  *ArrayEventContext*@10795}"EventContext: null"
      - typeCoercer = {$TypeCoercer_9660c5cdf3af@10797}""
      - values = {java.lang.Object[1]@10798}
      - values = null
```

Questions:
1. I assumed that the TypeCoercer can lookup ValueEncoder's. Check that this assumption is correct
2. Check if `ArrayEventContext` is invoking the `TypeCoercer`
3. Implement a custom `EventContext` if one of the core tapestry implementations does not suit (using URLEventContext as a guide) 

Discussion [here](http://tapestry.markmail.org/search/?q=list%3Aorg.apache.tapestry.users#query:list%3Aorg.apache.tapestry.users+page:1+mid:2ulkrv2jf5iycj6n+state:results)

**Comment by uklance** [comment]
Changes are probably required here:
https://github.com/uklance/tapestry-cometd/search?utf8=%E2%9C%93&q=ArrayEventContext

**Comment by ghost** [comment]
Hi Lance, as I am not that familiar with tapestry internals I may not be getting your question correctly, still here is my reply (if it helps)

1). When I have run a debug I could see that ArrayEventContext has TypeCoercer but in the list of

```
    typeToTargetCoercion = {java.util.WeakHashMap@11940} size = 21
    [0] = {java.util.AbstractMap$SimpleEntry@12382}"class java.lang.String" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12383}
    [1] = {java.util.AbstractMap$SimpleEntry@12384}"class org.lazan.t5.cometd.UpdateStrategy" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12385}
    [2] = {java.util.AbstractMap$SimpleEntry@12386}"class [Ljava.lang.Object;" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12387}
    [3] = {java.util.AbstractMap$SimpleEntry@12388}"class java.lang.Integer" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12389}
    [4] = {java.util.AbstractMap$SimpleEntry@12390}"interface org.apache.tapestry5.ValidationTracker" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12391}
    [5] = {java.util.AbstractMap$SimpleEntry@12392}"class domain.system.filter.BlackBoxFilterType" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12393}
    [6] = {java.util.AbstractMap$SimpleEntry@12394}"class domain.system.filter.BlackBoxFilter" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12395}
    [7] = {java.util.AbstractMap$SimpleEntry@12396}"class tapestry.domain.BlackBoxEditType" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12397}
    [8] = {java.util.AbstractMap$SimpleEntry@12398}"class java.lang.Object" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12399}
    [9] = {java.util.AbstractMap$SimpleEntry@12400}"class org.apache.tapestry5.corelib.data.BlankOption" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12401}
    [10] = {java.util.AbstractMap$SimpleEntry@12402}"class java.lang.Boolean" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12403}
    [11] = {java.util.AbstractMap$SimpleEntry@12404}"class java.lang.Long" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12405}
    [12] = {java.util.AbstractMap$SimpleEntry@12406}"class java.lang.Double" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12407}
    [13] = {java.util.AbstractMap$SimpleEntry@12408}"class org.apache.tapestry5.corelib.LoopFormState" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12409}
    [14] = {java.util.AbstractMap$SimpleEntry@12410}"class org.apache.tapestry5.ioc.util.TimeInterval" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12411}
    [15] = {java.util.AbstractMap$SimpleEntry@12412}"interface org.apache.tapestry5.ioc.Resource" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12413}
    [16] = {java.util.AbstractMap$SimpleEntry@12414}"class java.util.Date" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12415}
    [17] = {java.util.AbstractMap$SimpleEntry@12416}"class org.apache.tapestry5.corelib.data.InsertPosition" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12417}
    [18] = {java.util.AbstractMap$SimpleEntry@12418}"class tapestry.domain.ShotDataSection" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12419}
    [19] = {java.util.AbstractMap$SimpleEntry@12420}"interface org.apache.tapestry5.services.ValueEncoderFactory" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12421}
    [20] = {java.util.AbstractMap$SimpleEntry@12422}"class java.text.Format" -> {org.apache.tapestry5.ioc.internal.services.TypeCoercerImpl$TargetCoercion@12423}
```

I have also noticed that there is a certain event sequence that can force this to be populated with the Coercer for my objects, in which case this will work properly. However I still didn't figured out what that sequence is as it happened only once

2) Nope is it was unable to find it :)

Hope this helps, let me know if there is anything else I can/need to provide

**Comment by uklance** [comment]
Can you @Inject TypeCoercer into a page and try the following 

``` java
Coercion c = typeCoercer.getCoercion(String. class, Shot.class);
System.out. println(c);
```

**Comment by ghost** [comment]
That throws the exception

```
 org.apache.tapestry5.ioc.util.UnknownValueException
 Could not find a coercion from type java.lang.String to type domain.project.shot.Shot.
```

but I guess this is what you were looking for

**Comment by uklance** [comment]
Ah, I think I had my assumption the wrong way round. I thought TypeCoercer could pass through to ValueEncoder but it seems I had it the wrong way around (see [here](https://github.com/apache/tapestry5/search?utf8=%E2%9C%93&q=TypeCoercedValueEncoderFactory))

So the solution will be something like:

``` java
public class StringEventContext implements EventContext {
   private final ValueEncoderSource valueEncoderSource;
   private final String[] strings;

   public <T> T get(Class<T> desiredType, int index) {
      ValueEncoder<T> encoder =  valueEncoderSource.getValueEncoder(Class<T> type);
      return encoder.toValue(strings[index]);
   }

   // TODO: constructor and implement other methods
}
```

This new StringEventContext will replace ArrayEventContext usage [here](https://github.com/uklance/tapestry-cometd/blob/162db65e11fb976c69f06a9dca1f16f7cb0c2362/src/main/java/org/lazan/t5/cometd/services/internal/AuthorizersImpl.java#L99)

**Comment by ghost** [comment]
I see. OK when you push this new code maven let me know which version and I can pull in and test it out. Also if you know when that will happen I can see if I should implement some ugly workaround or can I just wait for the bug fix

Thanks for the help to identify this

**Comment by uklance** [comment]
Fixed in tapestry-cometd 0.9.19
